rthooks: support NRI

kkourt opened the pull request:
(Draft)

lambdanis commented on install/kubernetes/tetragon/templates/_container_rthooks.tpl:
```diff
@@ -16,6 +16,14 @@
     - --fail-allow-namespaces
     - {{ if .Values.rthooks.failAllowNamespaces }}{{ printf "%s,%s" .Release.Namespace .Values.rthooks.failAllowNamespaces }}{{ else }}{{ .Release.Namespace }}{{ end }}
+   {{- range $key, $value := .Values.rthooks.extraHookArgs }}
+   {{- if $value }}
```
`{{- if eq nil $value }}` would be better I think, to distinguish between `$value` being null or empty string/array/object.

kkourt replied:
Fixed, thanks!

lambdanis commented on contrib/tetragon-rthooks/cmd/oci-hook/main.go:
```diff
@@ -327,8 +327,11 @@ func main() {
 	switch cliConf.HookName {
 	case "createContainer":
+		// do nothing
+		log.Info("nothing to do", "hook", "createContainer")
```
Can we make it a Debug log?

kkourt replied:
I'd like to keep it as it is, if that's OK. If everything's correct, we should never hit this. And if we do, it's one message per container, which I think is fine.

lambdanis replied:
ok, if we're expecting to never hit this then yeah it's fine :)

lambdanis reviewed (commented):
I did one pass, left a few minor comments. I would need a second pass to fully follow the code logic, but looks good overall :)

kkourt commented on install/kubernetes/tetragon/templates/rthooks-daemonset.yaml:
```diff
@@ -0,0 +1,68 @@
+      annotations:
+        checksum/configmap: {{ toJson .Values.rthooks | sha256sum }}
+      {{- with .Values.rthooks.podAnnotations }}
+      annotations:
```
nice catch, thanks!

kkourt commented on install/kubernetes/tetragon/values.yaml:
```diff
@@ -205,6 +205,7 @@ tetragon:
   # does not run on the host.
   hostProcPath: "/proc"
   # -- Configure tetragon's init container for setting up tetragon-oci-hook on the host
+  # NOTE: This is deprecated, please use .rthooks
```
Yap, we should :)
Added one, thanks!

kkourt commented:
> I did one pass, left a few minor comments. I would need a second pass to fully follow the code logic, but looks good overall :)

Thanks Anna! I addressed the comments.

mtardy commented on contrib/tetragon-rthooks/cmd/setup/main.go:
```diff
@@ -24,16 +24,19 @@ const (
 	} `embed:"" prefix:"oci-hooks."`
+
+	HookArgs struct {
+		Args []string `arg:"" optional:""`
```
fyi: I had to discover kong to understand that `arg:""` would mean this is a positional argument and that all the rest would be added to this slice.

kkourt replied:
The trick is actually in the `passthrough:` of the HookArgs struct.

mtardy commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -0,0 +1,141 @@
+the topic.
+
+We use `minikube` as the example platform because it supports both `cri-o` and `containerd`. Also,
+at the time of this writing, no images that support this have been released, so we build images
+locally:
+
+```shell-session
+$ make image image-operator image-rthooks # build images within a tetragon checked out git repo
+```
````
````suggestion
at the time of this writing, no images that support this have been released, so we build images
locally within a tetragon checked out Git repo:

```shell
make image image-operator image-rthooks
```
````

mtardy commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -0,0 +1,141 @@
+
+### Setup Minikube
+
+```shell-session
+$ minikube start --driver=kvm2 --container-runtime=cri-o
+$ minikube image load cilium/tetragon:latest
+$ minikube image load cilium/tetragon-operator:latest
+$ minikube image load cilium/tetragon-rthooks:latest
+$ minikube image ls | grep tetragon
+localhost/cilium/tetragon:latest
+localhost/cilium/tetragon-rthooks:latest
+localhost/cilium/tetragon-operator:latest
+```
````
Following k8s.io convention and making the command copy-pastable:
````suggestion
```shell
minikube start --driver=kvm2 --container-runtime=cri-o
minikube image load cilium/tetragon:latest
minikube image load cilium/tetragon-operator:latest
minikube image load cilium/tetragon-rthooks:latest
minikube image ls | grep tetragon
```

The output should be similar to:

```
localhost/cilium/tetragon:latest
localhost/cilium/tetragon-rthooks:latest
localhost/cilium/tetragon-operator:latest
```
````

mtardy commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -0,0 +1,141 @@
+```
+### Install Tetragon
+
+```shell-session
+$ helm install \
+   --namespace kube-system \
+   --set tetragon.image.override=localhost/cilium/tetragon:latest \
+   --set tetragonOperator.image.override=localhost/cilium/tetragon-operator:latest \
+   --set rthooks.enabled=true \
+   --set rthooks.interface=oci-hooks \
+   --set rthooks.image.override=localhost/cilium/tetragon-rthooks:latest \
+   tetragon ./install/kubernetes/tetragon
+```
````
````suggestion
```shell
helm install \
   --namespace kube-system \
   --set tetragon.image.override=localhost/cilium/tetragon:latest \
   --set tetragonOperator.image.override=localhost/cilium/tetragon-operator:latest \
   --set rthooks.enabled=true \
   --set rthooks.interface=oci-hooks \
   --set rthooks.image.override=localhost/cilium/tetragon-rthooks:latest \
   tetragon ./install/kubernetes/tetragon
```
````

mtardy commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -0,0 +1,141 @@
+```
+
+
+```shell-session
+$ kubectl -n kube-system get pods | grep tetragon
+tetragon-hpjwq                       2/2     Running   0          2m42s
+tetragon-operator-664ddc8957-9lmd2   1/1     Running   0          2m42s
+tetragon-rthooks-m24xr               1/1     Running   0          2m42s
+```
````
````suggestion
```shell
kubectl -n kube-system get pods | grep tetragon
```

The output should be similar to:

```
tetragon-hpjwq                       2/2     Running   0          2m42s
tetragon-operator-664ddc8957-9lmd2   1/1     Running   0          2m42s
tetragon-rthooks-m24xr               1/1     Running   0          2m42s
```
````

mtardy commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -0,0 +1,141 @@
+### Test
+
+Start a pod:
+```shell-session
+$ kubectl run test --image=debian  --rm -it -- /bin/bash
+```
````
You can do the rest of fixes, basically not using `$` or `#` and [separate the cmd from the output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

mtardy commented on contrib/tetragon-rthooks/cmd/oci-hook/main.go:
```diff
@@ -173,7 +173,8 @@ func createContainerHook(log *slog.Logger) (error, map[string]string) {
 
 	if configName == "" {
-		return fmt.Errorf("failed to find spec file. Tried the following dirs: %v", configPaths), nil
+		log.Error("fail to find spec file", "paths", configPaths, "cwd", rootDir)
+		return fmt.Errorf("failed to find spec file"), nil
```
crazy nit: I guess you can factorize that error string into an error.New

mtardy commented on contrib/tetragon-rthooks/cmd/oci-hook/main.go:
```diff
@@ -368,3 +375,24 @@ func getBinaryDir() string {
+
+	connCtx, connCancel := context.WithTimeout(ctx, cliConf.GrpcTimeout)
+	defer connCancel()
+	conn, err := grpc.DialContext(connCtx, cliConf.AgentAddr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
```
Fyi: DialContext is deprecated and we need to move the codebase to the new way, since this is new code maybe you want to change it directly, reach out for more info!

kkourt replied:
OK, good to know. I added a commit to replace grpc.DialContext with grpc.NewClient.

mtardy reviewed (commented):
Nice patches! I just skipped this commit 4360fbf9f385d064a3dc7927d4fcd161c62afd3d to be honest but the rest looks good, here are some comments :)

kkourt commented on Dockerfile:
You mean `Dockerfile.rthooks`? Isn't it covered by: https://github.com/cilium/tetragon/blob/4ab963fad73904157f4b1456f9af5f9f1bfc6cd9/.github/renovate.json5#L40
?

mtardy replied:
No sorry, it’s confusing because I commented on the docker file but I meant line 38 and 39 of https://github.com/cilium/tetragon/blob/main/.github/renovate.json5Le

kkourt replied:
Ah, nice catch. Thanks, fixed.

kkourt commented on Dockerfile.rthooks:
```diff
@@ -0,0 +1,34 @@
+ARG TARGETOS
+# TARGETARCH is an automatic platform ARG enabled by Docker BuildKit.
+ARG TARGETARCH
+ARG NOSTRIP
```
yap, removed it.

mtardy commented:
No sorry, it’s confusing because I commented on the docker file but I meant line 38 and 39 of https://github.com/cilium/tetragon/blob/main/.github/renovate.json5Le 4 juil. 2024 à 14:04, Kornilios Kourtis ***@***.***> a écrit :﻿
@kkourt commented on this pull request.



On Dockerfile:
You mean Dockerfile.rthooks? Isn't it covered by: https://github.com/cilium/tetragon/blob/4ab963fad73904157f4b1456f9af5f9f1bfc6cd9/.github/renovate.json5#L40
?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because your review was requested.Message ID: ***@***.***>

kkourt commented on docs/content/en/docs/concepts/runtime-hooks/_index.md:
fixed.

kkourt commented on install/kubernetes/tetragon/values.yaml:
```diff
@@ -339,3 +340,52 @@ crds:
+
+# -- Method for installing Tetagon rthooks (tetragon-rthooks) daemonset
+# The tetragon-rthooks daemonset is responsible for installing run-time hooks on the host.
+# TODO: add documentation link
```
makes sense, added.

kkourt commented:
> Nice patches! I just skipped this commit [4360fbf](https://github.com/cilium/tetragon/commit/4360fbf9f385d064a3dc7927d4fcd161c62afd3d) to be honest but the rest looks good, here are some comments :)

Thanks for the review! Pushed a new version to address the comments.

mtardy reviewed (approved):
That's great! 🎉

tpapagian commented on docs/content/en/docs/installation/runtime-hooks.md:
````diff
@@ -76,3 +76,100 @@ Output:
+
+Install the new configuration file and restart containerd
+```shell
+cp /tmp/new-config.toml /etc/containerd/config.toml
````
Shouldn't we copy that inside minikube again? These commands seems to change the host `/etc/containerd/config.toml` file (and not the one inside minikube VM).

kkourt replied:
Nice catch! Yap, there is a `minikube` missing in front of the `cp`.

tpapagian reviewed (approved):
Thanks, this PR contains a lot of changes and it may be difficult to follow all of these during the review. But at a high level this LGTM!

Added also some minor comments.

kkourt commented on install/kubernetes/tetragon/values.yaml:
```diff
@@ -339,3 +340,52 @@ crds:
+  image:
+    override: ~
+    repository: quay.io/cilium/tetragon-rthooks
+    tag: v0.1
```
Nope, this should follow different versioning.
The reason is that we want to be able to upgrade the tetragon version without having to upgrade the hooks version. The plan for building the images is to add tags such as `rthooks/v0.1` to build the tetragon-rthooks images.

kkourt commented on docs/content/en/docs/concepts/runtime-hooks.md:
```diff
@@ -0,0 +1,59 @@
+information, the Tetragon agent can update the state so that Kubernetes Identify filtering can be
+applied in-kernel via BPF.
+
+One way that this information is available to the Tetragon agent is via the K8s API sever. Relying
```
thanks, fixed.

kkourt commented:
Merging this, thanks everyone for the helpful reviews!
